File: app/src/main/java/com/wqz/allinone/act_bookmark/LinkAddActivity.java
package com.wqz.allinone.act_bookmark;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.widget.EditText;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.wqz.allinone.R;
import com.wqz.allinone.database.BookmarkDBHelper;

import java.util.regex.Pattern;

public class LinkAddActivity extends AppCompatActivity {
    private EditText etName, etUrl;
    private BookmarkDBHelper bookmarkDBHelper;
    private int folderId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_link_add);

        Bundle bundle = getIntent().getExtras();
        assert bundle != null;
        folderId = bundle.getInt("folder_id");
        bookmarkDBHelper = new BookmarkDBHelper(this);

        etName = findViewById(R.id.et_name);
        etUrl = findViewById(R.id.et_url);
        findViewById(R.id.btn_add).setOnClickListener(v -> addLink());
    }

    private void addLink() {
        String name = etName.getText().toString().trim();
        String url = etUrl.getText().toString().trim();

        if (name.isEmpty() || url.isEmpty()) {
            Toast.makeText(this, "名称或链接不能为空", Toast.LENGTH_SHORT).show();
            return;
        }
        if (name.length() > 25 || name.length() < 2) {
            Toast.makeText(this, "名称长度不符合要求", Toast.LENGTH_SHORT).show();
            return;
        }
        if (name.contains("\n")) {
            Toast.makeText(this, "名称格式不正确", Toast.LENGTH_SHORT).show();
            return;
        }
        if (!isValidUrl(url)) {
            Toast.makeText(this, "链接格式不正确", Toast.LENGTH_SHORT).show();
            return;
        }

        bookmarkDBHelper.addLink(name, url, folderId);
        Toast.makeText(this, "添加成功", Toast.LENGTH_SHORT).show();
        Intent intent = new Intent();
        Bundle bundle = new Bundle();
        bundle.putInt("is_add", 1);
        intent.putExtras(bundle);
        setResult(Activity.RESULT_OK, intent);
        finish();
    }

    private boolean isValidUrl(String url) {
        String regex = "^(https?|ftp|file)://[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]";
        Pattern pattern = Pattern.compile(regex);
        return pattern.matcher(url).matches();
    }
}

File: app/src/main/java/com/wqz/allinone/act/bookmark/LinkAddActivity.kt
package com.wqz.allinone.act.bookmark

import android.content.Intent
import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.wqz.allinone.R
import com.wqz.allinone.act.bookmark.viewmodel.BookmarkViewModel
import com.wqz.allinone.entity.Link
import com.wqz.allinone.ui.XBackground
import com.wqz.allinone.ui.XCard
import com.wqz.allinone.ui.XItem
import com.wqz.allinone.ui.XToast
import com.wqz.allinone.ui.color.TextFieldColor
import java.util.regex.Pattern

/**
 * 链接添加
 * Created by Wu Qizhen on 2024.11.3
 * Refactored by Wu Qizhen on 2024.11.30
 */
class LinkAddActivity : ComponentActivity() {
    private lateinit var viewModel: BookmarkViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        viewModel = BookmarkViewModel(application)

        val folderId = intent.getIntExtra("FOLDER_ID", 0)

        setContent {
            XBackground.BreathingBackground(titleId = R.string.add_link) {
                LinkAddScreen(
                    folderId = folderId
                )
            }
        }
    }

    @Composable
    fun LinkAddScreen(
        folderId: Int = 0
    ) {
        val context = LocalContext.current
        var link by remember { mutableStateOf("") }
        var content by remember { mutableStateOf("") }

        XCard.SurfaceCard {
            TextField(
                modifier = Modifier.fillMaxWidth(),
                value = content,
                onValueChange = { content = it },
                colors = TextFieldColor.colors(),
                textStyle = TextStyle(
                    color = Color.White,
                    fontSize = 16.sp,
                    fontFamily = FontFamily(Font(R.font.misans_regular))
                ),
                label = {
                    Text(
                        text = "书签名",
                        fontWeight = FontWeight.Bold
                    )
                }
            )
        }

        Spacer(modifier = Modifier.height(10.dp))

        XCard.SurfaceCard {
            TextField(
                modifier = Modifier.fillMaxWidth(),
                value = link,
                onValueChange = { link = it },
                colors = TextFieldColor.colors(),
                textStyle = TextStyle(
                    color = Color.White,
                    fontSize = 16.sp,
                    fontFamily = FontFamily(Font(R.font.misans_regular))
                ),
                label = {
                    Text(
                        text = "URL",
                        fontWeight = FontWeight.Bold
                    )
                }
            )
        }

        Spacer(modifier = Modifier.height(10.dp))

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.Center
        ) {
            XItem.Button(icon = R.drawable.ic_visit, text = "批量") {
                val intent = Intent(this@LinkAddActivity, LinkImportActivity::class.java)
                intent.putExtra("FOLDER_ID", folderId)
                startActivity(intent)
                finish()
            }

            Spacer(modifier = Modifier.width(10.dp))

            XItem.Button(icon = R.drawable.ic_add, text = "添加") {
                val title = content.trim().replace("\n", "")
                val url = link.trim().replace("\n", "")
                if (title.isNotEmpty() && url.isNotEmpty()) {
                    if (isValidUrl(url)) {
                        val newLink = Link(
                            id = null,
                            title = title,
                            url = url,
                            folder = folderId
                        )
                        viewModel.insertLink(newLink)
                        XToast.showText(context, R.string.added)
                        finish()
                    } else {
                        XToast.showText(context, R.string.invalid_url)
                    }
                } else {
                    XToast.showText(context, R.string.input_link_empty)
                }
            }
        }

        Spacer(modifier = Modifier.height(50.dp))
    }

    private fun isValidUrl(url: String): Boolean {
        val regex = "^(https?|ftp|file)://[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]"
        val pattern = Pattern.compile(regex)
        return pattern.matcher(url).matches()
    }
}